docs/how-to/incremental-migrations.md

```diff
@@ -0,0 +1,365 @@
+
+This has several advantages:
+- SpacetimeDB's module hotswapping makes this a zero-downtime update. Write your new module, `spacetime publish` it, and watch the new table populate as it's used.
+- It amortizes the cost of transforming rows or computing new columns across many transactions. Rows will only be added to the new table when they're needed.
+- In many cases, old clients from before the update can coexist with new clients that use the new table. You can publish the updated module without disconnecting your clients, roll out the client update through normal channels, and allow your users to update at their own pace.
+
+For example, imagine we have a table `player` which stores information about our players:
+
+<!-- TODO: switchable language widget with C# version of below code. -->
+
+```rust
+#[spacetimedb::table(name = character, public)]
+pub struct Character {
+    #[primary_key]
+    #[auto_inc]
+    player_id: u64,
+    #[unique]
+    nickname: String,
+    level: u32,
+    class: Class,
+}
+
+#[derive(SpacetimeType, Debug, Copy, Clone)]
+pub enum Class {
+    Fighter,
+    Caster,
+    Medic,
+}
+```
+
+We'll write a few helper functions and some simple reducers:
+
+```rust
+#[spacetimedb::reducer]
+fn create_character(ctx: &ReducerContext, class: Class, nickname: String) {
+    log::info!(
+        "Creating new level 1 {class:?} named {nickname}",
+    );
+    ctx.db.character().insert(Character {
+        player_id: ctx.sender,
+        nickname,
+        level: 1,
+        class,
+    });
+}
+
+fn find_character_for_player(ctx: &ReducerContext) -> Character {
+    ctx.db
+        .character()
+        .player_id()
+        .find(ctx.sender)
+        .expect("Player has not created a character")
+}
+
+fn update_character(ctx: &ReducerContext, character: Character) {
+    ctx.db.character().player_id().update(character);
+}
+
+#[spacetimedb::reducer]
+fn rename_character(ctx: &ReducerContext, new_name: String) {
+    let character = find_character_for_player(ctx);
+    log::info!(
+        "Renaming {} to {}",
+        character.nickname,
+        new_name,
+    );
+    update_character(
+        ctx,
+        Character {
+            nickname: new_name,
+            ..character
+        },
+    );
+}
+
+#[spacetimedb::reducer]
+fn level_up_character(ctx: &ReducerContext) {
+    let character = find_character_for_player(ctx);
+    log::info!(
+        "Leveling up {} from {} to {}",
+        character.nickname,
+        character.level,
+        character.level + 1,
+    );
+    update_character(
+        ctx,
+        Character {
+            level: character.level + 1,
+            ..character
+        },
+    );
+}
+```
+
+We'll play around a bit with `spacetime call` to set up a character:
+
+```sh
+$ spacetime logs incr-migration-demo -f &
+
+$ spacetime call incr-migration-demo create_character '{ "Fighter": {} }' "Phoebe"
```

Comment:
Do we explain this enum syntax somewhere in the docs? That seems worth doing & linking to.

Reply:
It's in the JSON SATN ref. I'll add a link.